Bluetooth: Audio: Make PACS optional notify actually optional

fredrikdanebjer opened the pull request:
Added KConfig options to make the various PACS characteristics that are optionally notifiable selectable as notifiable. Selecting a characteristic as notifiable will make it so for all pacs instances on the device.

Fixes https://github.com/zephyrproject-rtos/zephyr/issues/56456

szymon-czapracki commented on subsys/bluetooth/audio/Kconfig.pacs:
```diff
@@ -83,9 +104,30 @@ config BT_PAC_SRC_LOC_WRITEABLE
+	  This option enables support for clients to be notified on the
+	  Supported Audio Contexts Characteristic.
+
+config BT_PACS_AVAILABLE_CONTEXT_NOTIFIABLE
```
Isn't notify in Available Contexts characteristics mandatory?

fredrikdanebjer replied:
Yes, you are right. That "None" camouflaged itself as a "Notify" very well for me

Thalley commented on subsys/bluetooth/audio/Kconfig.pacs:
```diff
@@ -12,6 +12,13 @@ config BT_PAC_SNK
 	help
 	  This option enables support for Sink Published Audio Capabilities.
 
+config BT_PAC_SNK_NOTIFIABLE
+	bool "Sink PAC Notifiable Support"
+	depends on BT_PAC_SNK
+	help
+	  This option enables support for clients to be notified of the Sink
+	  PAC Characteristic.
```
Suggest to move to under the `if BT_PAC_SNK` and remove `depends on BT_PAC_SNK`

Thalley commented on subsys/bluetooth/audio/Kconfig.pacs:
```diff
@@ -44,13 +51,27 @@ config BT_PAC_SNK_LOC_WRITEABLE
 	help
 	  This option enables support for Source Published Audio Capabilities.
 
+config BT_PAC_SRC_NOTIFIABLE
+	bool "Source PAC Notifiable Support"
+	depends on BT_PAC_SRC
+	help
+	  This option enables support for clients to be notified on the Source
+	  PAC Characteristic.
```
Suggest to move to under the `if BT_PAC_SRC` and remove `depends on BT_PAC_SRC`

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -550,36 +608,54 @@ BT_GATT_SERVICE_DEFINE(pacs_svc,
 	BT_AUDIO_CCC(src_cfg_changed),
 #if defined(CONFIG_BT_PAC_SRC_LOC)
-#if defined(CONFIG_BT_PAC_SRC_LOC_WRITEABLE)
+#if defined(CONFIG_BT_PAC_SRC_LOC_WRITEABLE) && defined(CONFIG_BT_PAC_SRC_LOC_NOTIFIABLE)
```
Can't comment on unmodified lines, but `BT_UUID_PACS_SRC` and `BT_UUID_PACS_SNK` are also optional notifiable.

Thalley reviewed (changes requested):
I expect this to heavily conflict with https://github.com/zephyrproject-rtos/zephyr/pull/55572. I suggest to get the https://github.com/zephyrproject-rtos/zephyr/pull/55572 in first, and then update this PR afterwards. 

This PR also partially fixes https://github.com/zephyrproject-rtos/zephyr/issues/57473, but does not contain support for subscribing in the client implementation.

fredrikdanebjer commented:
> I expect this to heavily conflict with #55572. I suggest to get the #55572 in first, and then update this PR afterwards.
> 
> This PR also partially fixes #57473, but does not contain support for subscribing in the client implementation.

Ill wait for this one and update once its in

Thalley commented:
@fredrikdanebjer This can now be rebased and be marked ready for review :)

fredrikdanebjer committed:
Bluetooth: Audio: Make PACS optional notify actually optional

Added Kconfig options to make the various PACS characteristics that are
optionally notifiable selectable as notifiable. Selecting a
characteristic as notifiable will make it so for all pacs instance on
the device.

Signed-off-by: Fredrik Danebjer <frdn@demant.com>

fredrikdanebjer commented:
@szymon-czapracki I'd urge you specifically to take a look at this patch, as it makes quite some changes to your recently merged PR https://github.com/zephyrproject-rtos/zephyr/pull/55572

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -666,92 +758,254 @@ static int pac_notify_snk(void)
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(clients); i++) {
+
+		for (uint8_t j = 0; j < FLAG_NUM; j++) {
+			if (atomic_test_bit(clients[i].flags, j)) {
+
+				/**
+				 *  It's enough that one flag is set, as the defer work will go
+				 * through all notifiable characteristics
+				 */
+				k_work_submit(&deferred_nfy_work);
+				return;
+			}
+		}
```
```suggestion
		if (atomic_get(&clients[i].flags) > 0) {
			/* It's enough that one flag is set, as the defer work will go through all 
			 * notifiable characteristics */
			k_work_submit(&deferred_nfy_work);
			break;
		}
```

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -666,92 +758,254 @@ static int pac_notify_snk(void)
+		return;
+	}
+
+	for (int i = 0; i < ARRAY_SIZE(clients); i++) {
```
```suggestion
	for (size_t i = 0; i < ARRAY_SIZE(clients); i++) {
```

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -666,92 +758,254 @@ static int pac_notify_snk(void)
+	}
+
+	if (!bt_addr_le_is_bonded(conn->id, &conn->le.dst)) {
+		LOG_DBG("2");
```
```suggestion
```

fredrikdanebjer replied:
Ouch, this was nasty :D

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -666,92 +758,254 @@ static int pac_notify_snk(void)
+	LOG_DBG("%s changed security level to %d", bt_addr_le_str(bt_conn_get_dst(conn)), level);
+
+	if (err != 0 || conn->encrypt == 0) {
+		LOG_DBG("1");
```
```suggestion
```

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -666,92 +758,254 @@ static int pac_notify_snk(void)
+
+static struct bt_conn_auth_info_cb auth_callbacks = {
+	.pairing_complete = pacs_auth_pairing_complete,
+	.bond_deleted = pacs_bond_deleted
```
Correct me if I am wrong, but I don't think `bond_delete` is called if a bond has been overwritten if `CONFIG_BT_KEYS_OVERWRITE_OLDEST=y`. 

You could argue that it should, but I don't think it currently is. I'll try to create a GH issue for that

Thalley replied:
Created https://github.com/zephyrproject-rtos/zephyr/issues/60239

fredrikdanebjer replied:
Is there any actions that should be taken to circumvent this feature/bug?

Thalley replied:
Seems there isn't any issue with this anymore

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -816,16 +1071,34 @@ int bt_pacs_cap_register(enum bt_audio_dir dir, struct bt_pacs_cap *cap)
 	}
 
+
+#if defined(CONFIG_BT_PAC_SNK_NOTIFIABLE)
+	pacs_set_notify_bit(FLAG_SINK_PAC_CHANGED);
+	bt_conn_foreach(BT_CONN_TYPE_LE, defer_value_ntf, NULL);
+
+#endif /* CONFIG_BT_PAC_SNK_NOTIFIABLE */
+
+#if defined(CONFIG_BT_PAC_SRC_NOTIFIABLE)
+	pacs_set_notify_bit(FLAG_SOURCE_PAC_CHANGED);
+	bt_conn_foreach(BT_CONN_TYPE_LE, defer_value_ntf, NULL);
+
+#endif /* CONFIG_BT_PAC_SRC_NOTIFIABLE */
+
```
```suggestion
#if defined(CONFIG_BT_PAC_SNK_NOTIFIABLE)
	if (dir == BT_AUDIO_DIR_SINK) {
		pacs_set_notify_bit(FLAG_SINK_PAC_CHANGED);
		k_work_submit(&deferred_nfy_work);
	}
#endif /* CONFIG_BT_PAC_SNK_NOTIFIABLE */

#if defined(CONFIG_BT_PAC_SRC_NOTIFIABLE)
	if (dir == BT_AUDIO_DIR_SOURCE) {
		pacs_set_notify_bit(FLAG_SOURCE_PAC_CHANGED);
		k_work_submit(&deferred_nfy_work);
	}
#endif /* CONFIG_BT_PAC_SRC_NOTIFIABLE */
```

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -849,7 +1122,11 @@ int bt_pacs_cap_unregister(enum bt_audio_dir dir, struct bt_pacs_cap *cap)
 	}
 
-	pacs_changed(dir);
+#if defined(CONFIG_BT_PAC_SRC_LOC_NOTIFIABLE)
+
+	pacs_set_notify_bit(FLAG_SOURCE_AUDIO_LOCATIONS_CHANGED);
+	bt_conn_foreach(BT_CONN_TYPE_LE, defer_value_ntf, NULL);
+#endif
```
Why are we notifying location here? 

Shouldn't we be notifying the PAC records?

fredrikdanebjer replied:
Yes, you are absolutely right. This is completely incorrect.

Thalley reviewed (changes requested):
I would also strongly suggest to update the babblesim tests to verify this new behavior

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -798,12 +1036,29 @@ void bt_pacs_cap_foreach(enum bt_audio_dir dir, bt_pacs_cap_foreach_func_t func,
+			LOG_DBG("Added %s to bonded list\n", addr_str);
+			return;
+		}
+	}
+
+}
```
```suggestion
	}
}
```

fredrikdanebjer commented:
> I would also strongly suggest to update the babblesim tests to verify this new behavior

This a good point. Ill take a look at this asap also

fredrikdanebjer committed:
Bluetooth: Audio: Make PACS notify bonded clients on reconnect

This fix allows PACS to notify bonded clients, that were disconnected
when a characteristic with notify property was changed, are being
reconnected.

This fix also implements a defer work so to not block the bt rx
thread during auth callbacks. All other notify calls are for
consistency and style also moved to use the defer work.

Signed-off-by: Fredrik Danebjer <frdn@demant.com>

fredrikdanebjer committed:
Bluetooth: Audio: Make BSIM tests for PACS notify functionality

Added a BSIM tests for PACS notification functionality. This test
tests that all optional notifies are working, as well as verifying
that a disconnected subscribed client will be notified on reconnect.

Signed-off-by: Fredrik Danebjer <frdn@demant.com>

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -275,19 +322,15 @@ static int set_supported_contexts(uint16_t contexts, uint16_t *supported,
-
-		return err;
-	}
+#if defined(CONFIG_BT_PACS_SUPPORTED_CONTEXT_NOTIFIABLE)
```
Can `IS_ENABLED` be used here instead?

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -63,15 +64,57 @@ static uint16_t src_available_contexts = BT_AUDIO_CONTEXT_TYPE_PROHIBITED;
 
+enum {
+	FLAG_ACTIVE,
+#if defined(CONFIG_BT_PAC_SNK_NOTIFIABLE)
+	FLAG_SINK_PAC_CHANGED,
+#endif /* CONFIG_BT_PAC_SNK_NOTIFIABLE) */
+#if defined(CONFIG_BT_PAC_SNK_LOC_NOTIFIABLE)
+	FLAG_SINK_AUDIO_LOCATIONS_CHANGED,
+#endif /* CONFIG_BT_PAC_SNK_LOC_NOTIFIABLE */
+#if defined(CONFIG_BT_PAC_SRC_NOTIFIABLE)
+	FLAG_SOURCE_PAC_CHANGED,
+#endif /* CONFIG_BT_PAC_SRC_NOTIFIABLE */
+#if defined(CONFIG_BT_PAC_SRC_LOC_NOTIFIABLE)
+	FLAG_SOURCE_AUDIO_LOCATIONS_CHANGED,
+#endif
+	FLAG_AVAILABLE_AUDIO_CONTEXT_CHANGED,
+#if defined(CONFIG_BT_PACS_SUPPORTED_CONTEXT_NOTIFIABLE)
+	FLAG_SUPPORTED_AUDIO_CONTEXT_CHANGED,
+#endif /* CONFIG_BT_PACS_SUPPORTED_CONTEXT_NOTIFIABLE */
```
Do we gain much from guarding these? 

Should `FLAG_AVAILABLE_AUDIO_CONTEXT_CHANGED` be guarded by `CONFIG_BT_PACS_SUPPORTED_CONTEXT_NOTIFIABLE`?

fredrikdanebjer replied:
> Do we gain much from guarding these?

I guess not very much. Since we already are guarding optional features like this, I prefer to have them, at least in my mind it gets a bit clearer.

> Should FLAG_AVAILABLE_AUDIO_CONTEXT_CHANGED be guarded by CONFIG_BT_PACS_SUPPORTED_CONTEXT_NOTIFIABLE?

No, because FLAG_AVAILABLE_AUDIO_CONTEXT_CHANGED  is not optional, and not dependent on supported contexts being notifiable or not.

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -355,29 +399,30 @@ static ssize_t snk_loc_read(struct bt_conn *conn,
-	err = pac_notify_snk_loc();
-
-	return err;
+#if defined(CONFIG_BT_PAC_SNK_LOC_NOTIFIABLE)
```
Ditto with IS_ENABLED

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -482,37 +520,30 @@ static ssize_t src_loc_read(struct bt_conn *conn,
-	}
-
-	return 0;
+#if defined(CONFIG_BT_PAC_SRC_LOC_NOTIFIABLE)
```
Ditto

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -666,92 +757,233 @@ static int pac_notify_snk(void)
+		return;
+	}
+
+#if defined(CONFIG_BT_PAC_SNK_NOTIFIABLE)
+	if (atomic_test_and_clear_bit(client->flags, FLAG_SINK_PAC_CHANGED)) {
+		LOG_DBG("Notifying Sink PAC");
+		pac_notify(conn, BT_AUDIO_DIR_SINK);
+	}
+#endif /* CONFIG_BT_PAC_SNK_NOTIFIABLE) */
+#if defined(CONFIG_BT_PAC_SNK_LOC_NOTIFIABLE)
+	if (atomic_test_and_clear_bit(client->flags, FLAG_SINK_AUDIO_LOCATIONS_CHANGED)) {
+		LOG_DBG("Notifying Sink Audio Location");
+		pac_notify_loc(conn, BT_AUDIO_DIR_SINK);
+	}
+#endif /* CONFIG_BT_PAC_SNK_LOC_NOTIFIABLE */
+#if defined(CONFIG_BT_PAC_SRC_NOTIFIABLE)
+	if (atomic_test_and_clear_bit(client->flags, FLAG_SOURCE_PAC_CHANGED)) {
+		LOG_DBG("Notifying Source PAC");
+		pac_notify(conn, BT_AUDIO_DIR_SOURCE);
+	}
+#endif /* CONFIG_BT_PAC_SRC_NOTIFIABLE */
+#if defined(CONFIG_BT_PAC_SRC_LOC_NOTIFIABLE)
+	if (atomic_test_and_clear_bit(client->flags, FLAG_SOURCE_AUDIO_LOCATIONS_CHANGED)) {
+		LOG_DBG("Notifying Source Audio Location");
+		pac_notify_loc(conn, BT_AUDIO_DIR_SOURCE);
+	}
+#endif
+	if (atomic_test_and_clear_bit(client->flags, FLAG_AVAILABLE_AUDIO_CONTEXT_CHANGED)) {
+		LOG_DBG("Notifying Available Contexts");
+		available_contexts_notify(conn);
+	}
+#if defined(CONFIG_BT_PACS_SUPPORTED_CONTEXT_NOTIFIABLE)
+	if (atomic_test_and_clear_bit(client->flags, FLAG_SUPPORTED_AUDIO_CONTEXT_CHANGED)) {
+		LOG_DBG("Notifying Supported Contexts");
+		supported_contexts_notify(conn);
+	}
+#endif /* CONFIG_BT_PACS_SUPPORTED_CONTEXT_NOTIFIABLE */
```
If we don't guard the flags, we could use IS_ENABLED here

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -666,92 +757,233 @@ static int pac_notify_snk(void)
+		return;
+	}
+
+	for (size_t i = 0U; i < ARRAY_SIZE(clients); i++) {
+
+		for (size_t j = 0U; j < FLAG_NUM; j++) {
```
```suggestion
	for (size_t i = 0U; i < ARRAY_SIZE(clients); i++) {
		for (size_t j = 0U; j < FLAG_NUM; j++) {
```

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -816,16 +1048,38 @@ int bt_pacs_cap_register(enum bt_audio_dir dir, struct bt_pacs_cap *cap)
+	}
+
+
+#if defined(CONFIG_BT_PAC_SNK_NOTIFIABLE)
```
Ditto with IS_ENABLED

Thalley commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -816,16 +1048,38 @@ int bt_pacs_cap_register(enum bt_audio_dir dir, struct bt_pacs_cap *cap)
 
+#endif /* CONFIG_BT_PAC_SNK_NOTIFIABLE */
+
+#if defined(CONFIG_BT_PAC_SRC_NOTIFIABLE)
```
Ditto with IS_ENABLED

Thalley commented on tests/bsim/bluetooth/audio/src/pacs_notify_client_test.c:
```diff
@@ -0,0 +1,580 @@
+	SET_FLAG(flag_is_encrypted);
+}
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
```
Please don't use `BT_CONN_CB_DEFINE` for the audio Babblesim tests, as this will always be registered regardless of whether we are running the PACS tests. Consider registering dynamically if you cannot use the common `connected` callback

Thalley commented on tests/bsim/bluetooth/audio/src/pacs_notify_client_test.c:
```diff
@@ -0,0 +1,580 @@
+{
+	printk("Number of notifications received: %d\n", num_notifications++);
+
+	if (num_notifications >= 6) {
```
Suggest to make this more explicit and test for whether you received all the different notifications, rather than just a total number.

Thalley commented on tests/bsim/bluetooth/audio/src/pacs_notify_client_test.c:
```diff
@@ -0,0 +1,580 @@
+	pacs_instance.discover_params.end_handle = BT_ATT_LAST_ATTRIBUTE_HANDLE;
+	pacs_instance.discover_params.type = BT_GATT_DISCOVER_CHARACTERISTIC;
+	pacs_instance.discover_params.func = discover_pacs_src;
+
+
```
```suggestion

```

Thalley commented on tests/bsim/bluetooth/audio/src/pacs_notify_client_test.c:
```diff
@@ -0,0 +1,580 @@
+static uint8_t discover_pacs_snk(struct bt_conn *conn, const struct bt_gatt_attr *attr,
+			     struct bt_gatt_discover_params *params)
+{
+	struct bt_gatt_subscribe_params *subscribe_params;
+
+	int err;
```
```suggestion
	struct bt_gatt_subscribe_params *subscribe_params;
	int err;
```

Thalley commented on tests/bsim/bluetooth/audio/src/pacs_notify_client_test.c:
```diff
@@ -0,0 +1,580 @@
+
+static struct pacs_instance_t pacs_instance;
+
+static void security_changed(struct bt_conn *conn, bt_security_t level, enum bt_security_err err)
+{
+	if (err) {
+		FAIL("Failed to change security (err %d)\n", err);
+		return;
+	}
+
+	if (level != BT_SECURITY_L2) {
+		FAIL("Wrong security level set (level %d)", level);
+		return;
+	}
+
+	SET_FLAG(flag_is_encrypted);
+}
```
You could probably skip this. The GATT layer in Zephyr automatically increases security when needed

Thalley commented on tests/bsim/bluetooth/audio/src/pacs_notify_server_test.c:
```diff
@@ -0,0 +1,236 @@
+	int err;
+	const struct bt_data ad[] = {
+		BT_DATA_BYTES(BT_DATA_FLAGS, (BT_LE_AD_GENERAL | BT_LE_AD_NO_BREDR)),
+	};
+	printk("Enabling Bluetooth\n");
```
```suggestion
	};
	
	printk("Enabling Bluetooth\n");
```

MariuszSkamra commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -627,92 +730,249 @@ static int pac_notify_snk(void)
+	};
+	int err;
+
+	if (!IS_ENABLED(CONFIG_BT_PACS_SUPPORTED_CONTEXT_NOTIFIABLE)) {
+		return -ENOTSUP;
+	}
```
This check seems to be redundant, as the only call is guarded with:
```
	if (IS_ENABLED(CONFIG_BT_PACS_SUPPORTED_CONTEXT_NOTIFIABLE) &&
	    atomic_test_bit(client->flags, FLAG_SUPPORTED_AUDIO_CONTEXT_CHANGED)) {
```

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.pacs:
```diff
@@ -44,6 +50,13 @@ config BT_PAC_SNK_LOC_WRITEABLE
+	bool "Sink Audio Location Notifiable Support"
+	depends on BT_PAC_SNK_LOC
+	help
+	  This option enables support for clients to be notified on the Sink
+	  Audio Location Characteristic.
```
```suggestion
	  This option enables support for clients to be notified on the Sink
	  Audio Location changes.
```

fredrikdanebjer replied:
For all of these I prefer then to change it to 'Audio Location Characteristic changes.', since we everywhere else have it mentioned explictely as a characteristic

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.pacs:
```diff
@@ -31,6 +31,12 @@ config BT_PACS_SNK_CONTEXT
+config BT_PAC_SNK_NOTIFIABLE
+	bool "Sink PAC Notifiable Support"
+	help
+	  This option enables support for clients to be notified of the Sink
+	  PAC Characteristic.
```
```suggestion
	  This option enables support for clients to be notified on the Sink
	  PAC changes.
```

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.pacs:
```diff
@@ -70,6 +83,12 @@ config BT_PACS_SRC_CONTEXT
+config BT_PAC_SRC_NOTIFIABLE
+	bool "Source PAC Notifiable Support"
+	help
+	  This option enables support for clients to be notified on the Source
+	  PAC Characteristic.
```
```suggestion
	  This option enables support for clients to be notified on the Source
	  PAC changes.
```

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.pacs:
```diff
@@ -83,9 +102,23 @@ config BT_PAC_SRC_LOC_WRITEABLE
+	bool "Source Audio Location Notifiable Support"
+	depends on BT_PAC_SRC_LOC
+	help
+	  This option enables support for clients to be notified on the Source
+	  Audio Location Characteristic.
```
```suggestion
	  This option enables support for clients to be notified on the Source
	  Audio Location changes.
```

MariuszSkamra commented on subsys/bluetooth/audio/Kconfig.pacs:
```diff
@@ -83,9 +102,23 @@ config BT_PAC_SRC_LOC_WRITEABLE
+	bool "Supported Audio Context Notifiable Support"
+	depends on BT_PACS
+	help
+	  This option enables support for clients to be notified on the
+	  Supported Audio Contexts Characteristic.
```
```suggestion
	  This option enables support for clients to be notified on the
	  Supported Audio Contexts changes.
```

MariuszSkamra commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -499,96 +519,157 @@ static ssize_t src_loc_write(struct bt_conn *conn,
 #if defined(CONFIG_BT_PAC_SNK)
+#if defined(CONFIG_BT_PAC_SNK_NOTIFIABLE)
 	BT_AUDIO_CHRC(BT_UUID_PACS_SNK,
 		      BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY,
 		      BT_GATT_PERM_READ_ENCRYPT,
 		      snk_read, NULL, NULL),
 	BT_AUDIO_CCC(snk_cfg_changed),
+#else
+	BT_AUDIO_CHRC(BT_UUID_PACS_SNK,
+		      BT_GATT_CHRC_READ,
+		      BT_GATT_PERM_READ_ENCRYPT,
+		      snk_read, NULL, NULL),
+#endif /* CONFIG_BT_PAC_SNK_NOTIFIABLE */
 #if defined(CONFIG_BT_PAC_SNK_LOC)
-#if defined(CONFIG_BT_PAC_SNK_LOC_WRITEABLE)
+#if defined(CONFIG_BT_PAC_SNK_LOC_WRITEABLE) && defined(CONFIG_BT_PAC_SNK_LOC_NOTIFIABLE)
 	BT_AUDIO_CHRC(BT_UUID_PACS_SNK_LOC,
 		      BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE | BT_GATT_CHRC_NOTIFY,
 		      BT_GATT_PERM_READ_ENCRYPT | BT_GATT_PERM_WRITE_ENCRYPT,
 		      snk_loc_read, snk_loc_write, NULL),
-#else
+	BT_AUDIO_CCC(snk_loc_cfg_changed),
+#elif defined(CONFIG_BT_PAC_SNK_LOC_NOTIFIABLE)
 	BT_AUDIO_CHRC(BT_UUID_PACS_SNK_LOC,
 		      BT_GATT_CHRC_READ | BT_GATT_CHRC_NOTIFY,
 		      BT_GATT_PERM_READ_ENCRYPT,
 		      snk_loc_read, NULL, NULL),
-#endif /* CONFIG_BT_PAC_SNK_LOC_WRITEABLE */
 	BT_AUDIO_CCC(snk_loc_cfg_changed),
+#elif defined(CONFIG_BT_PAC_SNK_LOC_WRITEABLE)
+	BT_AUDIO_CHRC(BT_UUID_PACS_SNK_LOC,
+		      BT_GATT_CHRC_READ | BT_GATT_CHRC_WRITE,
+		      BT_GATT_PERM_READ_ENCRYPT,
+		      snk_loc_read, snk_loc_write, NULL),
+#else
+	BT_AUDIO_CHRC(BT_UUID_PACS_SNK_LOC,
+		      BT_GATT_CHRC_READ,
+		      BT_GATT_PERM_READ_ENCRYPT,
+		      snk_loc_read, NULL, NULL),
+#endif /* (CONFIG_BT_PAC_SNK_LOC_WRITEABLE && CONFIG_BT_PAC_SNK_LOC_NOTIFIABLE) */
```
It's a bit hard to read (and maintain) IMO. Maybe consider using some macros e.g.
```
#define BT_PACS_SNK_LOC_PERM \
        BT_GATT_CHRC_READ \
        IF_ENABLED(CONFIG_BT_PAC_SNK_LOC_WRITEABLE, (| BT_GATT_CHRC_WRITE), ())
#define BT_PACS_SNK_LOC_PROP \
        BT_GATT_CHRC_READ \
        IF_ENABLED(CONFIG_BT_PAC_SNK_LOC_WRITEABLE, (| BT_GATT_CHRC_WRITE), ()) \
        IF_ENABLED(CONFIG_BT_PAC_SNK_LOC_NOTIFIABLE, (| BT_GATT_CHRC_NOTIFY), ())
#define BT_PACS_SNK_LOC(_read, _write) \
	BT_AUDIO_CHRC(BT_UUID_PACS_SNK_LOC, BT_PACS_SNK_LOC_PROP, \
                      BT_PACS_SNK_LOC_PERM, _read, _write, NULL), \
        IF_ENABLED(CONFIG_BT_PAC_SNK_LOC_NOTIFIABLE, (BT_AUDIO_CCC(snk_loc_cfg_changed),), ())
```

applies to all characteristics

MariuszSkamra replied:
Let's leave it as it is. We can refactor this later.

MariuszSkamra replied:
I admit I was wrong, what's your opinion?

fredrikdanebjer replied:
I have already made this change :D Please take a look.
I think it turns out to be more readable indeed.

MariuszSkamra commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -597,12 +678,29 @@ static int pac_notify_src_loc(void)
+	struct bt_uuid *uuid;
+
+	switch (dir) {
+	case BT_AUDIO_DIR_SINK:
+		if (!IS_ENABLED(CONFIG_BT_PAC_SNK_NOTIFIABLE)) {
+			return -ENOTSUP;
+		}
```
As it's internal function, called by stack I would make an assert here

MariuszSkamra commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -627,92 +730,249 @@ static int pac_notify_snk(void)
-#if defined(CONFIG_BT_PAC_SRC)
-		err = pac_notify_src();
-#endif /* CONFIG_BT_PAC_SRC */
+	LOG_DBG("Notify cb");
```
```suggestion
	LOG_DBG("");
```

MariuszSkamra commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -627,92 +730,249 @@ static int pac_notify_snk(void)
+
+	if (IS_ENABLED(CONFIG_BT_PAC_SNK_NOTIFIABLE) &&
+	    atomic_test_bit(client->flags, FLAG_SINK_PAC_CHANGED)) {
+		printk("Notifying Sink PAC\n");
```
Avoid using `printk`, use logging subsystem (LOG_DBG) instead. Applies in multiple places

fredrikdanebjer replied:
Yes, this was indeed some nasty leftover :)

fredrikdanebjer commented on subsys/bluetooth/audio/pacs.c:
```diff
@@ -23,6 +23,7 @@
 #include <zephyr/bluetooth/audio/pacs.h>
 #include <zephyr/sys/slist.h>
 #include "../host/conn_internal.h"
+#include "../host/hci_core.h"
```
I do because of usage of `bt_addr_le_is_bonded`
